### PR TITLE
Dynamic registration of RPC users security roles in ArtemisMessagingS…

### PR DESCRIPTION
**Pull request by IgorNitto**
Change ArtemisMessagingServer security management in a way that does not require knowledge of RPC usernames at startup.  Dynamic registration of RPC users is an implicit requirement in the future scenario where authentication data will be  loaded from an external service (see [ENT-1000]).

Note: added Mike, as this touches security critical code. This approach relies on the SecuritySettingPlugin features of Artermis. Ideas for alternative implementations and suggestions on testing are very welcome.

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt [hunk 1]
```diff
@@ -676,3 +686,51 @@ class NodeLoginModule : LoginModule {
...
+private class RolesAdderOnLogin(val source: (String) -> Pair<String, Set<Role>>)
+    : SecuritySettingPlugin, LoginListener {
+
+    /*
+     * Artemis internal container storing roles association
+     */
+    private lateinit var repository: RolesRepository
+
+    override fun onLogin(username: String) {
+        val (address, roles) = source(username)
```
Comment (IgorNitto) [inline node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt]: RolesRepository should already be thread-safe, however I could explicitly make this synchronized for extra safety

**Comment by sollecitom** [comment]
Great job! @mikehearn and @mnesbit please have a look because it's quite sensitive code.

**Comment by sollecitom** [comment]
Only think to mention is that this will "document" a reference to an enterprise branch into open source. Not a big deal if you ask me.

**Comment by sollecitom** [comment]
Matthew has basically agreed to the change, even if he pointed out this area of the code might change at some point.

**Comment by mnesbit** [comment]
I am fine with the intent of this and I will approve now. However, I regard this as abandon-ware to meet a deadline. We really should not depend upon any server side hooks for authentication as this ties us firmly to In-Process Artemis and no other broker, nor does it mesh well with clustering. I would far prefer to see design patterns that were based upon abstract messaging patterns e.g. challenge-response messages to get an RPC session queue that is auto-deleted on disconnect and app level tracking of live clients sessions.

**Comment by sollecitom** [comment]
This PR is just got early on preview. We have a standing one including
these changes. Can we have it working this way for now (it's essentially as
it was working before) so that we can deliver users read from a database,
and create a ticket about changing the RPC client as you said? We can work
on the ticket but I don't think we have time to rework this in scope for we
15th.

File: node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt [hunk 2]
```diff
@@ -555,6 +563,7 @@ class NodeLoginModule : LoginModule {
         this.subject = subject
         this.callbackHandler = callbackHandler
         userService = options[RPCUserService::class.java.name] as RPCUserService
+	    loginListener = options[LoginListener::class.java.name] as LoginListener
```
[the remark this answers is not in the document]
Comment (IgorNitto) [reply node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt]: Yes, I am using intelliJ but for some reason it is hiding these spaces.

**Comment by IgorNitto** [comment]
Thanks for reviewing. I understand it does not go in the direction of "decoupling" from Artemis, which is bad and not acceptable in the long run. I will be happy to contribute to a more stable solution as the architecture of the underlying messaging system evolves.

**Pull request merged**